### ISSUE TITLE
GRPO loss-path metrics average micro-batch entries, so uneven accumulation windows are weighted by their size

### ISSUE THREAD

**Issue by behroozazarkhalili** [huggingface/trl#7047]
### Summary

`GRPOTrainer` appends its loss-path metrics (`policy_loss`, `entropy`, and every metric scaled by `current_gradient_accumulation_steps` in `_compute_loss`) once per micro-batch, and `log()` takes a flat mean over the entries of the logging interval. When the interval contains accumulation windows of different sizes, that mean weights each window by its micro-batch count instead of counting each optimizer step once. The logged value is then a different number from the optimizer-window average, while the base `loss` reported by transformers is unaffected.

### Where the uneven windows come from

transformers shortens the last window of an epoch to the remainder and sets `current_gradient_accumulation_steps` to that size (`transformers/trainer.py`, 5.11.0):

```python
remainder = steps_in_epoch % self.args.gradient_accumulation_steps
if remainder == 0:
    remainder = self.args.gradient_accumulation_steps
...
num_batches = (
    self.args.gradient_accumulation_steps if update_step != (num_update_steps_per_epoch - 1) else remainder
)
...
self.current_gradient_accumulation_steps = len(batch_samples)
```

So with `gradient_accumulation_steps=3` and an epoch of four micro-batches the windows are three micro-batches then one. Each micro-batch entry in `self._metrics["train"]` is already scaled to its own window (`accumulation_factor = self.current_gradient_accumulation_steps`), so the entries carry the right values. The bias comes from the count: the long window contributes three entries and the short one contributes one, and `log()` divides by four.

### Measured on main (5dd51e4d, transformers 5.11.0, CPU)

`trl-internal-testing/tiny-Qwen2ForCausalLM-2.5`, four prompts from `trl-internal-testing/zen`, `per_device_train_batch_size=2`, `gradient_accumulation_steps=3`, `steps_per_generation=2`, `num_generations=2`, `max_completion_length=4`, `entropy_coef=0.1`, `logging_strategy="epoch"`, a deterministic reward. The instrumented trainer records each micro-batch's `current_gradient_accumulation_steps` and the entry appended to `_metrics`.

| micro-batch | window size | `policy_loss` entry |
| --- | --- | --- |
| 1 | 3 | 0.0 |
| 2 | 3 | 0.0 |
| 3 | 3 | 0.0 |
| 4 | 1 | -0.0022285581 |

| aggregation | `policy_loss` |
| --- | --- |
| logged (flat mean over 4 entries) | -0.00055713951587677 |
| mean over the 2 optimizer windows | -0.00111427903175354 |

The logged value is half the per-window mean, which is exactly the three-to-one entry weighting. `entropy` is appended unscaled (`global_masked_mean(entropies)`), so it only sees the count weighting, not the window-size scaling: 11.929378747940063 logged versus 11.929380893707275 per window, a gap that is small because the four micro-batch entries are nearly identical. The base `loss` does not have the problem: transformers sums the scaled micro-batch losses and divides by the optimizer-step count, and its logged value matches the per-window mean to float rounding (-1.194052338600159 versus -1.194052368402481).

### Scope

Every entry on this path has the property: the normalizer branches in `_compute_loss` (around lines 3244 to 3269 on main), the entropy accumulation factor (around 3287), and the `policy_loss` capture. GMPO, MiniLLM and the replay-buffer trainer inherit `_compute_loss`, so they log the same numbers. #7042 corrects the scale of one `policy_loss` branch to match the others and deliberately leaves the aggregation alone, so this issue is the follow-up.

### Suggested fix

Accumulate each metric's micro-batch contributions into a per-window sum and append to `self._metrics` once, at the window boundary (`self._step % self.current_gradient_accumulation_steps == 0` already exists in `training_step`), so the entry count equals the optimizer-step count and `log()`'s mean is exact. Training is untouched; only the logged values change, and only when an interval mixes window sizes.

I can open the PR.

**Comment by adityasharmaaaaa** [comment]
Solid writeup, the entry-count vs optimizer-step-count distinction is easy to miss. If you don't get to opening the PR or want a second pair of eyes on GMPO/MiniLLM/replay-buffer coverage once the fix lands, happy to help.

**Comment by behroozazarkhalili** [comment]
Opened #7053 with the per-window aggregation: each metric accumulates its micro-batch contributions and is appended once per optimizer window at the gradient-sync boundary. Re-running the measurement above against that branch logs `policy_loss` -0.00111427903175354, the mean over the two windows, with the training loss unchanged.

**Comment by behroozazarkhalili** [comment]
Thanks. The PR is open as #7053. It changes `GRPOTrainer` and the two experimental copies that duplicate `_compute_loss` (GMPO and GSPO-token); MiniLLM and the replay-buffer trainer subclass `GRPOTrainer` without overriding the loss or logging path, so they pick the change up through inheritance. A second read on #7053 is welcome, especially on whether any other experimental trainer reimplements the `_metrics[mode]` append.

**Comment by adityasharmaaaaa** [comment]
Went through `main` and grepped all references to `current_gradient_accumulation_steps` to check for the same pattern: a metric scaled by the accumulation factor, appended once per micro-batch, and then flat-averaged in `log()`.

I found only two other instances of this pattern, and both are already covered by this PR:

* `experimental/gmpo/gmpo_trainer.py` — uses the same `normalizer = self.current_gradient_accumulation_steps if mode == "train" else 1.0` for `kl` / `entropy` / `clip_ratio/*`.
* `experimental/gspo_token/grpo_trainer.py` — has a full `_compute_loss` reimplementation with the same scaling applied to the same metrics.

`MiniLLMTrainer` and `GRPOWithReplayBufferTrainer` don't override `_compute_loss` or `log()`, so they inherit the fix as noted in the PR.

For the remaining references to `current_gradient_accumulation_steps` (`rloo_trainer.py`, core `distillation_trainer.py`, `sdft`, `sdpo`, `ssd`, `async_grpo`, `async_distillation`, and `server_distillation`), the factor is only applied to the returned loss for backward, or—in the RLOO/distillation cases—to a `step_time` metric that is already gated on the accumulation-window boundary (`self._step % self.current_gradient_accumulation_steps == 0`).

None of those scale a `_metrics[...]` value before appending, so there isn't a corresponding scale/count mismatch to fix.

One tangential thing I noticed while checking `async_grpo_trainer.py`: `ratio` / `kl` / `entropy` / `clip_ratio/*` are appended once per micro-batch and reduced via a flat `sum / len` in `_reduce_metric`. With uneven accumulation windows, this means they're still an equal-weight-per-microbatch mean rather than a per-optimizer-step mean.

Not related to this bug and not blocking this PR, but flagging it in case it's worth a follow-up issue.

**Comment by behroozazarkhalili** [comment]
Thanks, that matches my own grep of main. The sdft, sdpo and ssd sites divide only the returned loss (ssd appends its cross-entropy metric before the division), the RLOO and distillation sites gate a `step_time` on the window boundary, and the async trainers divide the loss only, so none of them has the scale-then-flat-mean mismatch.

On the async_grpo tangent: I read it the same way, but I would not file it. Those metrics are appended unscaled once per micro-batch and `_reduce_metric` takes a plain mean, so with uneven windows the result is an equal-weight-per-micro-batch mean rather than a per-optimizer-step mean. That is a different estimator, not a biased one: nothing is multiplied by a window size that the divisor then ignores, which is what made the GRPO number wrong. Whether a per-micro-batch or per-window mean is the better summary is a reporting choice, and the async trainer's own docstring on `_reduce_metric` reads as a deliberate one.

**Comment by adityasharmaaaaa** [comment]
Makes sense, the distinction holds up. In GRPO, the value is scaled by window size and then divided by the entry count, so those factors don’t cancel out when window sizes differ. That’s the actual bug.

In async_grpo, nothing is scaled going in, so taking a flat mean over micro-batches is a consistent (albeit different) estimator rather than a biased one. Agreed that it’s not worth a follow-up issue.

I’ll take a pass at #7053 — mainly checking that the GMPO/GSPO-token `_metrics[mode]` append sites line up with the windowed accumulation, and confirming that MiniLLM/replay-buffer don’t shadow `_compute_loss` or `log()` anywhere.

**Comment by qgallouedec** [comment]
logged entropy 11.929378747940063 vs per-window 11.929380893707275 is a 1.8e-7 relative gap. Entries scaled by 3 and by 1 would differ by ~2x no?

**Comment by behroozazarkhalili** [comment]
You are right, and the entropy row in the issue is mis-stated. The `entropy` metric is appended unscaled (`global_masked_mean(entropies)`, grpo_trainer.py:3356 on main); the `accumulation_factor` at line 3284 scales only the entropy bonus inside the loss, not the logged value. So the 1.8e-7 gap is the pure count-weighting effect (three near-identical entries versus one), not a 3x-vs-1x scale mismatch. `policy_loss` is the metric on the scaled path and is where the 2x shows. I have corrected the issue text; #7053's per-window accumulation is the same fix for both kinds of entry.

**Comment by behroozazarkhalili** [comment]
Closing this myself. I found it by reading the code, not from a failure in a real training run, and per #6862 that isn't the bar here. Happy to reopen if someone hits it in practice.
